# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В
свободное время жители городов больше не выходят на улицу, не посещают кафе
и торговые центры. Зато стало больше времени для книг. Это заметили
стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг
по подписке. Ваша первая задача как аналитика — проанализировать базу
данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские
обзоры книг. Эти данные помогут сформулировать ценностное предложение для
нового продукта.

## Описание таблиц


Таблица `books` — содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.
  
Таблица `authors` — содержит данные об авторах:  
- author_id — идентификатор автора;  
- author — имя автора.  
  
Таблица `publishers` — содержит данные об издательствах:  
- publisher_id — идентификатор издательства;  
- publisher — название издательства;  
  
Таблица `ratings` — содержит данные о пользовательских оценках книг:  
- rating_id — идентификатор оценки;  
- book_id — идентификатор книги;  
- username — имя пользователя, оставившего оценку;  
- rating — оценка книги.  
  
Таблица `reviews` — содержит данные о пользовательских обзорах на книги:   
- review_id — идентификатор обзора;  
- book_id — идентификатор книги;  
- username — имя пользователя, написавшего обзор;  
- text — текст обзора.  


## Откроем БД

In [ ]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas


def sql_query(query):
    result = pd.io.sql.read_sql(query, con=engine)
    return result


In [ ]:
query = '''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

'''

sql_query(query)

schemaname           tablename         tableowner tablespace  hasindexes  \
0      public              orders    praktikum_admin       None        True   
1      public              visits    praktikum_admin       None        True   
2      public  advertisment_costs    praktikum_admin       None        True   
3      public             authors    praktikum_admin       None        True   
4      public              second  praktikum_student       None       False   
5      public            second_b  praktikum_student       None       False   
6      public          publishers    praktikum_admin       None        True   
7      public              author  praktikum_student       None        True   
8      public             reviews    praktikum_admin       None        True   
9      public             ratings    praktikum_admin       None        True   
10     public               books    praktikum_admin       None        True   

    hasrules  hastriggers  rowsecurity  
0      False        False        False  
1      False        False        False  
2      False        False        False  
3      False         True        False  
4      False        False        False  
5      False        False        False  
6      False         True        False  
7      False        False        False  
8      False         True        False  
9      False         True        False  
10     False         True        False

In [ ]:
for x in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    print('\n', x)
    query = f'''

    SELECT *
    FROM {x}
    LIMIT 5;

    '''

    display(sql_query(query))


 books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


 authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


 publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


 ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


 reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [ ]:
# типы данных
query = '''
SELECT
    table_name,
    column_name,
    data_type,
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
'''

sql_query(query)

table_name       column_name data_type is_nullable
0      authors         author_id   integer          NO
1      authors            author      text         YES
2   publishers      publisher_id   integer          NO
3   publishers         publisher      text         YES
4      reviews         review_id   integer          NO
5      reviews           book_id   integer         YES
6      reviews          username      text         YES
7      reviews              text      text         YES
8      ratings         rating_id   integer          NO
9      ratings           book_id   integer         YES
10     ratings          username      text         YES
11     ratings            rating   integer         YES
12       books           book_id   integer          NO
13       books         author_id   integer         YES
14       books             title      text         YES
15       books         num_pages   integer         YES
16       books  publication_date      date         YES
17       books      publisher_id   integer         YES

## Решение задач

**Задача 1**

**Посчитайте, сколько книг вышло после 1 января 2000 года;**

In [ ]:
query = '''

SELECT COUNT(title) AS total_books
FROM books
WHERE publication_date >= '2000-01-01';

'''

sql_query(query)

total_books
0          821

**Комментарий**

Вышла 821 книга

**Задача 2**

**Для каждой книги посчитайте количество обзоров и среднюю оценку;**

In [ ]:
query = '''

WITH b AS (SELECT title, book_id
        FROM books),
i AS (SELECT AVG(rating) AS avg_rating, book_id
        FROM ratings
        GROUP BY book_id),
s AS (SELECT COUNT(review_id) AS total_reviews, book_id
        FROM reviews
        GROUP BY book_id)
SELECT  b.title, i.avg_rating, s.total_reviews
FROM b
LEFT OUTER JOIN i ON b.book_id=i.book_id
LEFT OUTER JOIN s ON b.book_id=s.book_id
-- WHERE s.total_reviews IS NOT NULL
GROUP BY b.title, s.total_reviews, i.avg_rating
ORDER BY s.total_reviews DESC;

'''

sql_query(query)#.head(10)

title  avg_rating  \
0      Anne Rice's The Vampire Lestat: A Graphic Novel    3.666667   
1    Disney's Beauty and the Beast (A Little Golden...    4.000000   
2                            Essential Tales and Poems    4.000000   
3                                 Leonardo's Notebooks    4.000000   
4     The Cat in the Hat and Other Dr. Seuss Favorites    5.000000   
..                                                 ...         ...   
995                                The Social Contract    3.500000   
996  To Green Angel Tower (Memory  Sorrow  and Thor...    4.500000   
997               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1    2.500000   
998                              Wicked: The Grimmerie    3.500000   
999                  Winter Prey (Lucas Davenport  #5)    4.500000   

     total_reviews  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
..             ...  
995            1.0  
996            1.0  
997            1.0  
998            1.0  
999            1.0  

[1000 rows x 3 columns]

**Комментарий**


Расчитала среднее кол-во обзоров и оценок, для 6 книг оценок нет

**Задача 3**


**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;**




In [ ]:
query = '''

SELECT b.publisher_id, p.publisher, COUNT(b.book_id) AS total_books
FROM books AS b
LEFT OUTER JOIN publishers AS p ON b.publisher_id=p.publisher_id
WHERE num_pages >= 50
GROUP BY b.publisher_id, p.publisher
ORDER BY total_books DESC
LIMIT 1;

'''

sql_query(query)

publisher_id      publisher  total_books
0           212  Penguin Books           42

**Комментарий**

Penguin Books

**Задача 4**


**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;**


In [ ]:
query = '''

SELECT a.author, ROUND(AVG(rating), 2) AS score
FROM books AS b
LEFT OUTER JOIN authors AS a ON b.author_id=a.author_id
LEFT OUTER JOIN ratings AS ra ON b.book_id=ra.book_id
WHERE b.book_id IN (SELECT book_id
                  FROM ratings
                  GROUP BY book_id
                  HAVING COUNT(rating) >= 50)
GROUP BY a.author
ORDER BY score DESC
LIMIT 1

'''

sql_query(query)

author  score
0  J.K. Rowling/Mary GrandPré   4.29

**Комментарий**

J.K. Rowling/Mary GrandPré с оценкой 4.29

**Задача 5**

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [ ]:
query = '''

WITH t AS (SELECT r.username, count(r.review_id) as count
FROM reviews AS r
WHERE r.username IN (SELECT ra.username
                    FROM ratings AS ra
                    GROUP BY ra.username
                    HAVING COUNT(rating) > 48)
GROUP BY r.username)

SELECT avg(count)
FROM t

'''

sql_query(query)

avg
0  24.0

**Комментарий**

среднее кол-во обзоров 24

## Вывод



1 Всего 821 книга вышла после 1 января 2000 года

2 Для каждой из 1000 книг посчитала средний рейтинг и количество обзоров. Для 6 из 1000 отсутствуют обзоры.

3 Penguin Books выпустило наибольшее количество книг — 42.

4 Самая высокая средняя оценка у J.K. Rowling/Mary GrandPré — 4.29.

5 Посчитала количество обзоров от пользователей которые поставили больше 48 оценок.